### PR TITLE
Add Low-Level Networking

### PR DESCRIPTION
**Pull request by Erdroy**
This PR adds low-level networking based upon NetworkManager and INetworkDriver, that allows easy transport-layer replacement (ENet, Valve Sockets etc.).

Default network driver implementation is using ENet.

Draft until ready.

### CODE REVIEW

**Comment by mafiesto4** [comment]
C# API added (<https://github.com/FlaxEngine/FlaxEngine/commit/5e401b6a788909ec77c53d5f6b6869b834464b39>) so we can move forward with this.

**Comment by Erdroy** [comment]
Marking as ready for review. We have to test this out, make some fixes (if needed) - SAMPLE NEEDED! 
+ docs pages would be awesome, but this will come after the sample is ready.

Idea for the sample: Simple chat that allows multiple clients join to server via ipv4 and port. Basic UI with chat box, client sends message to the server, server sends it to all other clients.

File: Source/Engine/Networking/NetworkMessage.cpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) 2012-2021 Wojciech Figat. All rights reserved.
+
+#include "NetworkMessage.h"
+
+void NetworkMessage::WriteBytes(uint8* bytes, const int numBytes)
```
Comment (mafiesto4) [inline Source/Engine/Networking/NetworkMessage.cpp:5]: Maybe move `WriteBytes` into `NetworkMessage.h`?

File: Source/Engine/Networking/NetworkMessage.cpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+{
+    ASSERT(Position + numBytes < BufferSize);
+    Platform::MemoryCopy(Buffer + Position, bytes, numBytes);
+    Position += numBytes;
+}
+
+void NetworkMessage::ReadBytes(uint8* bytes, const int numBytes)
```
Comment (mafiesto4) [inline Source/Engine/Networking/NetworkMessage.cpp:12]: Maybe move `ReadBytes` into `NetworkMessage.h`?

File: Source/Engine/Networking/NetworkMessage.cpp [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+{
+    ASSERT(Position + numBytes < BufferSize);
+    Platform::MemoryCopy(bytes, Buffer + Position, numBytes);
+    Position += numBytes;
+}
+
+bool NetworkMessage::IsValid() const
```
Comment (mafiesto4) [inline Source/Engine/Networking/NetworkMessage.cpp:19]: Maybe move `IsValid` into `NetworkMessage.h`?

File: Source/Engine/Networking/NetworkPeer.cpp [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+#include "Drivers/ENetDriver.h"
+
+#include "Engine/Core/Log.h"
+#include "Engine/Core/Math/Math.h"
+#include "Engine/Platform/CPUInfo.h"
+
+namespace
+{
+    Array<NetworkPeer*, HeapAllocation> Peers;
+    uint32_t LastHostId = 0;
```
Comment (mafiesto4) [inline Source/Engine/Networking/NetworkPeer.cpp:15]: `uint32`  🍔

File: Source/Engine/Networking/NetworkPeer.h [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    /// <param name="channelType">The channel to send the message over.</param>
+    /// <param name="message">The message.</param>
+    /// <param name="targets">The connections list to send the message to.</param>
+    /// <remarks>Can be used only by the server!</remarks>
+    /// <remarks>
+    /// Do not recycle the message after calling this.
+    /// This function automatically recycles the message.
+    /// </remarks>
+    API_FUNCTION()
+    bool EndSendMessage(NetworkChannelType channelType, const NetworkMessage& message, Array<NetworkConnection, HeapAllocation> targets);
```
Comment (mafiesto4) [inline Source/Engine/Networking/NetworkPeer.h:152]: Pass `targets` as reference to reduce memory allocations.

File: Source/Engine/Networking/INetworkDriver.h [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    /// </summary>
+    /// <param name="channelType">The channel to send the message over.</param>
+    /// <param name="message">The message.</param>
+    /// <param name="targets">The connections list to send the message to.</param>
+    /// <remarks>Can be used only by the server!</remarks>
+    /// <remarks>
+    /// Do not recycle the message after calling this.
+    /// This function automatically recycles the message.
+    /// </remarks>
+    virtual void SendMessage(NetworkChannelType channelType, const NetworkMessage& message, Array<NetworkConnection, HeapAllocation> targets) = 0;
```
Comment (mafiesto4) [inline Source/Engine/Networking/INetworkDriver.h:96]: Pass `targets` as reference to reduce memory allocations.

**Review by mafiesto4 — CHANGES_REQUESTED** [review]
Looks good! I left some comments for improvements. After merging we will need to fix compilation issues on PS4/UWP/XboxOne (or use those platforms online subsystems as network drivers or implement raw driver that runs purely over low-level platform sockets).

File: Source/ThirdParty/enet/enet.Build.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+/// https://github.com/zpl-c/enet
+/// </summary>
+public class ENet : DepsModule
+{
+    /// <inheritdoc />
+    public override void Init()
+    {
+        base.Init();
+
+        LicenseType = LicenseTypes.Custom;
```
[the remark this answers is not in the document]
Comment (Erdroy) [reply Source/ThirdParty/enet/enet.Build.cs:17]: Done

File: Source/Engine/Networking/Drivers/ENetDriver.cpp [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+            eventPtr->EventType = NetworkEventType::Connected;
+
+            if(IsServer())
+                _peerMap.Add(connectionId, event.peer);
+            break;
+            
+        case ENET_EVENT_TYPE_DISCONNECT:
+            eventPtr->EventType = NetworkEventType::Disconnected;
+            
+            if(IsServer() && _peerMap.ContainsKey(connectionId))
```
[the remark this answers is not in the document]
Comment (Erdroy) [reply Source/Engine/Networking/Drivers/ENetDriver.cpp]: Done

**Review by mafiesto4 — APPROVED** [review]

**Pull request merged**